contracts/src/gateway/router/XnetMessagingFacet.sol

```diff
@@ -29,4 +29,28 @@ contract XnetMessagingFacet is GatewayActorModifiers {
     function applyCrossMessages(IpcEnvelope[] calldata crossMsgs) external systemActorOnly {
         LibGateway.applyMessages(s.networkName.getParentSubnet(), crossMsgs);
     }
+
+    /// @notice Applies top-down cross-net message locally one by one.
+    ///         This is invoked by IPC nodes when drawing messages from their parent
+    ///         subnet for local execution. That's why the sender is restricted to the system sender,
+    ///         because this method is implicitly invoked by the node during block production.
+    /// @dev It requires the caller to be the system actor.
+    /// @param crossMsg The cross-network message to be applied.
+    function applyMsgWithRet(
+        IpcEnvelope calldata crossMsg
+    ) external systemActorOnly returns (bool hasReturn, ApplyMsgRet memory) {
+        return LibGateway.applyMsgWithRet(s.networkName.getParentSubnet(), crossMsg);
+    }
+
+    /// @notice Sends a receipt from the execution of a cross-message.
+    /// Only `Call` messages trigger a receipt. Transfer messages should be directly
+    /// handled by the peer client to return the funds to the from address in the
+    /// failing network. This should be only called for top-down message execution.
+    function sendReceipt(
+        IpcEnvelope calldata originalCrossMsg,
+        bool success,
+        bytes memory ret
+    ) external systemActorOnly {
+        LibGateway.sendReceipt(originalCrossMsg, success, ret);
```

Comment:
that's the original implementation, but that wont catch the vm errors, which is not handled by the contract.

Reply:
What's a VM error, why does it need catching? If the FVM fails, we don't want to send receipts anyway AFAIU. 

Here's how I thought the Solidity method would work:
* We have a single Solidity method to execute a whole batch of messages (or just one with a MMR proof, whatever)
* The Solidity method does the delegation or whatever trick it needs to do to actually process the message and gets the result. 
* It emits an event about which message was executed, to make it easy for clients to monitor the execution and outcome of cross messages.
* It can also insert an ack message if that's how you want to send results back. 

This way you don't have to aggregate multiple `ApplyRet` into a single one to sum up the gas costs, nor do you have to accumulate events: there is 1 call to the FVM all succeeding or failing together on the VM level (eg. if the relayer itself cannot cover the gas cost), with as many events about the outcomes as there are messages in the batch. 

You may or may not want to count the gas cost of the application of finality.